### PR TITLE
feat(`forge`) `_expectRevertCheatcode`

### PR DESCRIPTION
**Pull request by klkvr**
## Motivation

Ref #6833 

Considering that we need a way to write negative tests for cheatcodes internally but making existing `expectRevert` cheats interact with cheatcodes as well would break a lot of thing and not really a desired feature, this PR adds `_expectRevertCheatcode` cheats for internal usage.

## Solution

1. Extends `ExpectedRevert` context to allow 2 kinds of expectations - `Default` (expects revert from next non-cheatcode call) and `Cheatcode` (expects revert from next cheatcode call)
2. This cheatcode shouldn't end up in forge-std, so I've marked it as `Experimental`.
3. Updated cheatcode tests to use `_expectRevertCheatcode` instead of `expectRevert`

The logic with `pending` bool might be a little bit confusing, so maybe there is a way to improve it which I haven't figured out yet. I tried to document it clearly

### CODE REVIEW

**Comment by klkvr** [comment]
Failing CI is triggered by one of the tests being false-positive earlier and failing on Windows now, not sure how to fix such things :/

File: crates/cheatcodes/src/test/expect.rs [hunk 1]
```diff
@@ -53,12 +53,26 @@ pub enum ExpectedCallType {
     Count,
 }
 
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
+pub enum ExpectedRevertType {
```
Comment (mattsse) [inline crates/cheatcodes/src/test/expect.rs:57]: needs one line doc

Comment (mattsse) [inline crates/cheatcodes/src/test/expect.rs:57]:
```suggestion
#[derive(Clone, Debug)]
pub enum ExpectedRevertKind {
```

File: crates/cheatcodes/src/test/expect.rs [hunk 2]
```diff
@@ -53,12 +53,26 @@ pub enum ExpectedCallType {
...
+}
+
+#[derive(Clone, Debug)]
 pub struct ExpectedRevert {
     /// The expected data returned by the revert, None being any
     pub reason: Option<Vec<u8>>,
     /// The depth at which the revert is expected
     pub depth: u64,
+    /// The type of expected revert.
+    pub revert_type: ExpectedRevertType,
```
Comment (mattsse) [inline crates/cheatcodes/src/test/expect.rs:75]:
```suggestion
    pub kind: ExpectedRevertKind,
```

**Review by mattsse — CHANGES_REQUESTED** [review]
given that this behaviour is somewhat required to move forward with converting forge-std checks to rust native cheatcodes, I think we should proceed with this.

having an additional cheatcode for internal use makes sense to me, but we probably don't want to expose this via the forge-std interface. Perhaps there's an easy way to exclude them when we generate the interface, maybe with an additional group (`Internal`?)

wdyt @DaniPopes

File: crates/cheatcodes/src/inspector.rs [hunk 3]
```diff
@@ -919,61 +920,83 @@ impl<DB: DatabaseExt> Inspector<DB> for Cheatcodes {
...
+            call.contract == CHEATCODE_ADDRESS || call.contract == HARDHAT_CONSOLE_ADDRESS;
 
-        if data.journaled_state.depth() == 0 && self.skip {
-            return (
-                InstructionResult::Revert,
-                remaining_gas,
-                super::Error::from(MAGIC_SKIP).abi_encode().into(),
-            );
-        }
+        if !cheatcode_call {
```
Comment (Evalir) [inline crates/cheatcodes/src/inspector.rs:926]: let's add docs on how we expect the flow to go

**Review by Evalir — CHANGES_REQUESTED** [review]
Looking good so far, and agreed we need to proceed with this to make these assertions native. About naming, maybe `expectCheatcodeRevert` is clearer?

**Comment by klkvr** [comment]
@Evalir agreed that `expectCheatcodeRevert` is better, renamed it and pushed other fixes

pending guidance on whether we want `Internal` cheats group here

File: crates/cheatcodes/spec/src/vm.rs [hunk 4]
```diff
@@ -680,6 +680,18 @@ interface Vm {
     #[cheatcode(group = Testing, safety = Unsafe)]
     function expectRevert(bytes calldata revertData) external;
 
+    /// Expects an error on next cheatcode call with any revert data.
+    #[cheatcode(group = Testing, safety = Unsafe, status = Experimental)]
```
Comment (mattsse) [inline crates/cheatcodes/spec/src/vm.rs:684]: yeah I think adding an Internal variant would be appropriate

File: testdata/cheats/Fs.t.sol [hunk 5]
```diff
@@ -23,10 +23,10 @@ contract FsTest is DSTest {
 
         assertEq(vm.readFile(path), "hello readable world\nthis is the second line!");
 
-        vm.expectRevert(FOUNDRY_READ_ERR);
+        vm._expectCheatcodeRevert(FOUNDRY_READ_ERR);
```
Comment (mattsse) [inline testdata/cheats/Fs.t.sol:26]: can we still use expectRevert here, or would this no longer be possible?

Comment (klkvr) [reply testdata/cheats/Fs.t.sol:26]: We can, but the test will exit and pass right after the reverted call

That is the current behavior which resulted in some false positives

File: crates/cheatcodes/src/test/expect.rs [hunk 6]
```diff
@@ -53,12 +53,27 @@ pub enum ExpectedCallType {
...
+#[derive(Clone, Debug)]
+pub enum ExpectedRevertKind {
+    /// Expects revert from the next non-cheatcode call.
+    Default,
+    /// Expects revert from the next cheatcode call.
+    ///
+    /// The `pending` flag is used to track whether we have exited `expectCheatcodeRevert` context
+    /// or not. We have to track it to avoid expecting `expectCheatcodeRevert` call to revert
+    /// itself.
+    Cheatcode { pending: bool },
```
Comment (mattsse) [inline crates/cheatcodes/src/test/expect.rs:66]: I'd like a more expressive name for `pending` here.
perhaps simply `pending_processing` or even flip to `processed: bool`

**Review by mattsse — CHANGES_REQUESTED** [review]
changes lgtm,
pendatic naming nit.

Could you please also summarize what the impact of this change is wrt regular expectRevert?

**Comment by klkvr** [comment]
> Could you please also summarize what the impact of this change is wrt regular expectRevert?

This PR will not affect existing `expectRevert` behavior, it simply changes flow in `call_end` hook making it a bit more explicit about how cheatcode calls are being treated.

**Comment by mattsse** [comment]
pending @Evalir @DaniPopes

**Comment by klkvr** [comment]
> we'd def figure out a way to not include this in the vm interface

forge-std script to create Vm.sol interface already excludes `Experimental` cheats, so I think we can just exclude `Internal` as well

https://github.com/foundry-rs/forge-std/blob/ae570fec082bfe1c1f45b0acca4a2b4f84d345ce/scripts/vm.py#L32-L34

**Review by mattsse — APPROVED** [review]
lgtm

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**